### PR TITLE
Copy custom extension flags in a call to SSL_set_SSL_CTX()

### LINKED ISSUE #2180: Calling SSL_set_SSL_CTX loses supported client TLS extensions
It's common to use SNI and call SSL_set_SSL_CTX with a new context once the host name has been negotiated, for example the nginx webserver does this (https://github.com/nginx/nginx/blob/f8a9d528df92c7634088e575e5c3d63a1d4ab8ea/src/http/ngx_http_request.c#L888)

Unfortunately SSL_set_SSL_CTX replaces the the certificate (ssl->cert, https://github.com/openssl/openssl/blob/7d152a3c4f58bc2c4e94946e2a8dbf1a855a52f2/ssl/ssl_lib.c#L3484), which also contains the client extensions from the ClientHello in ssl->cert->cli_ext. So when an extension later checks for client support, nothing is found (https://github.com/openssl/openssl/blob/7d152a3c4f58bc2c4e94946e2a8dbf1a855a52f2/ssl/t1_ext.c#L44).

This breaks things like the certificate transparency extension which needs to be negotiated after SNI (see https://github.com/grahamedgecombe/nginx-ct/issues/13 for an affected use case).

### PR DESCRIPTION
**Pull request by mattcaswell**
The function SSL_set_SSL_CTX() can be used to swap the SSL_CTX used for
a connection as part of an SNI callback. One result of this is that the
s->cert structure is replaced. However this structure contains information
about any custom extensions that have been loaded. In particular flags are
set indicating whether a particular extension has been received in the
ClientHello. By replacing the s->cert structure we lose the custom
extension flag values, and it appears as if a client has not sent those
extensions.

SSL_set_SSL_CTX() should copy any flags for custom extensions that appear
in both the old and the new cert structure.

Fixes #2180 

This version is for master. Versions for 1.1.0 and 1.0.2 coming up.

##### Checklist

- [ ] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by mattcaswell** [comment]
Also see #3426 for the 1.1.0 version and #3427 for the 1.0.2 version.

File: ssl/statem/extensions_cust.c [hunk 1]
```diff
@@ -231,6 +231,26 @@ int custom_ext_add(SSL *s, int context, WPACKET *pkt, X509 *x, size_t chainidx,
...
 }
 
+/* Copy the flags from src to dst for any extensions that exist in both */
+int custom_exts_copy_flags(custom_ext_methods *dst,
+                           const custom_ext_methods *src)
+{
+    size_t i;
+
+    for (i = 0; i < src->meths_count; i++) {
+        custom_ext_method *methsrc = src->meths + i;
```
Comment (richsalz) [inline ssl/statem/extensions_cust.c]: Make this a local variable and put it in the for loop

Comment (mattcaswell) [inline ssl/statem/extensions_cust.c]: Done

**Comment by mattcaswell** [comment]
Updated commit pushed addressing review feedback. Similar commits made and pushed to #3426 and #3427.

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**